Introduces a quota subsystem

bom-d-van opened the pull request:
The quota subsystem is made to improve control, reliability and visibility of go-carbon.
It is not a standard graphite component, but it is backward compatible and
could be turned on optionally.

Caveat: the current implementation only supports concurrent/realtime trie index.

The quota subsystem allows user to control how many resources can be consumed
on a patter-matching based basis.

Implemented controls include: data points (based retention policy), disk size
(logical and physical), throughput, metric count, and namespaces (i.e. immediate
sub-directory count).

**More details could be found in [doc/quotas.md](https://github.com/go-graphite/go-carbon/blob/quota/doc/quotas.md) in the PR.**

An example configuration:

```ini
# This control all the namespaces under root
[*]
metrics       =       1,000,000
logical_size  = 250,000,000,000
physical_size =  50,000,000,000
# max means practically no limit
data_points   =             max
throughput    =             max

[sys.app.*]
metrics       =         3,000,000
logical_size  = 1,500,000,000,000
physical_size =   100,000,000,000
data_points   =   130,000,000,000

# This controls the root/global limits
[/]
namespaces    =                20
metrics       =        10,000,000
logical_size  = 2,500,000,000,000
physical_size = 2,500,000,000,000
data_points   =   200,000,000,000
dropping_policy = new
```

Throttling control is implemented in `carbonserver`, while quota config 
is implemented in persister (mainly for convenience).

bom-d-van committed:
Introduces a quota subsystem

The quota subsystem is made to improve control, reliability and visibility of go-carbon.
It is not a standard graphite component, but it is backward compatible and
could be turned on optionally.

Caveat: the current implementation only supports concurrent/realtime trie index.

The quota subsystem allows user to control how many resources can be consumed
on a patter-matching based basis.

Implemented controls include: data points (based retention policy), disk size
(logical and physical), throughput, metric count, and namespaces (i.e. immediate
sub-directory count).

More details could be found in doc/quotas.md in the PR.

An example configuration:

```ini
[*]
metrics       =       1,000,000
logical_size  = 250,000,000,000
physical_size =  50,000,000,000
data_points   =             max
throughput    =             max

[sys.app.*]
metrics       =         3,000,000
logical_size  = 1,500,000,000,000
physical_size =   100,000,000,000
data_points   =   130,000,000,000

[/]
namespaces    =                20
metrics       =        10,000,000
logical_size  = 2,500,000,000,000
physical_size = 2,500,000,000,000
data_points   =   200,000,000,000
dropping_policy = new
```

Throttling control is implemented in `carbonserver`, while quota config
is implemented in persister (mainly for convenience).

grzkv commented on doc/quotas.md:
```diff
@@ -0,0 +1,110 @@
+
+## Available controls
+
+* namespaces: how many sub-namespaces are allowed for the matched prefix/namespaces
```
@bom-d-van It is not entirely clear to me what use-case does this cover. Why do we need this?

bom-d-van replied:
For namespaces, it can be used to limit the number of top namespaces allowed to create in the cluster/instance.
For example, limit the number of namespaces under `sys.` to 10 give go-carbon administrator control to avoid users pollute the top metric namespaces.

grzkv commented on doc/quotas.md:
```diff
@@ -0,0 +1,110 @@
+
+* The current implementation is based on concurrent/realtime trie-index, so to use quotas, also need to enable those features on go-carbon.
+* Pattern matching rules aren't regexp, but like graphite query syntax, using globs
+* Quota and usage metrics are only produced if
```
I guess, there'll be a continuation :)

grzkv commented on doc/quotas.md:
```diff
@@ -0,0 +1,110 @@
+* metrics: how many metrics are allowed for the matched prefix/namespaces
+* logical-size: how many logical disk space are allowed for the matched prefix/namespaces
+* physical-size: how many physical disk space are allowed for the matched prefix/namespaces
+* data-points: how many data points (inferred using storage-schemas.conf) are allowed for the matched prefix/namespaces
```
@bom-d-van Do we really need this kind of fine-graining? As above, which use-cases does this cover?

bom-d-van replied:
there are more explanation bellow about data-points and logical-size, there are some overlaps between two of them.
like other control options listed there, could be enabled if needed.

grzkv commented on deploy/storage-quota.conf:
```diff
@@ -0,0 +1,24 @@
+# This control all the namespaces under root
+[*]
+metrics       =       1,000,000
+logical_size  = 250,000,000,000
```
Are commas required for the number or optional?

bom-d-van replied:
optional, added for a quick and small improvement of readability.

bom-d-van committed:
persister: simplifies PathError in logs

> simplifyPathError retruns an error without path in the error message. This
> simpplies the log a bit as the path is usually printed separately and the
> new error message is easier to filter in elastic search and other tools.

bom-d-van committed:
carbonserver: generate knonwMetrics by counting trie index files

WHY: with concurrent and realtime index, disk scan should be set at
am interval like 2 hours or longer. couting the files in trie index
gives us more timely visibilitty on how many metrics are known now.

bom-d-van committed:
quota: reduce dir node allocation in throttle by pre-allocation

bom-d-van committed:
quota: should only flush quotaAndUsageMetrics.quotaAndUsageMetrics once

The original implementation would keep resending the quota and usage metrics after the
initial flush, which is not helpful. Using a channel to avoid duplicate flushes.

bom-d-van committed:
quota: adopt a more performant throughput throttling implementation

When it comes to point search, map lookup is very fast, much faster than traversing the trie tree.
What's more, the current trie implementation adopts a space efficient strategy and children
are not sorted. And this means much higher cpu usage. We are seeing up to 700% cpu usage growth
in a cluster receving 1 millions+ data points.

By adopting a map-based implementation, we are able to cut down up to 600% of cpu usage in
that cluster, which means it is much more efficient to collect and control throughput usage.

bom-d-van committed:
quota: resolve deepsource warnings/errors

bom-d-van commented:
Hi guys, these changes are finally ready for review (took a bit of force-pushes to get deepsource happy). Can you take a look when you have time and let me know if we can get it merged?

The changes are deployed on prod for a few weeks and not more new issues are identified for now. As always, it's intended to be backward-compatible and should have no breaking changes.

cc @deniszh @azhiltsov @Civil

azhiltsov commented:
How many stat metrics should I expect with this in Quota config? 
```
[*]
``` 
or with this
```
[sys.app.*]
```
As many as quantity of metrics inside, or this is not going to resolve the * to the actual names?
What will be the naming scheme for those metrics?

bom-d-van commented:
> How many stat metrics should I expect with this in Quota config?
> 
> ```ini
> [*]
> ```
> 
> or with this
> 
> ```
> [sys.app.*]
> ```
> 
> As many as quantity of metrics inside, or this is not going to resolve the * to the actual names?
> What will be the naming scheme for those metrics?

Hi @azhiltsov , it's a like writing a graphite query. For example, supposed that there are 3 top level namespaces in the go-carbon instance: `sys`, `user`, `net`, the `*` is the same as writing separate 3 quota rules with the same config:

```ini
[*]
metrics = 1,000,000
```

the config above is the same as the config bellow for the example:

```ini
[user]
metrics = 1,000,000

[sys]
metrics = 1,000,000

[net]
metrics = 1,000,000
```

Same for `sys.app.*`, it matches the namespaces under `sys.app`.

Hope this  explains it.

bom-d-van committed:
quota: fix test bug and deepsource lint error due to origin/master rebase

bom-d-van commented:
Hi everyone, the subsystem has been on our production for sometime and it appears to be stable.

If no objection, I will proceed to merge the changes today. Feel free to let me know that if you have concerns.

deniszh commented:
Yes, looks good, @bom-d-van , please proceed.
Will make new release soon after merging this too, we have bunch of changes thanks to you!

bom-d-van commented:
Thank you Denis. Merging it.
